src/windows/process.rs

```diff
@@ -530,18 +527,68 @@ unsafe fn get_region_size(handle: HANDLE, ptr: LPVOID) -> Result<usize, &'static
     Ok((meminfo.RegionSize as isize - ptr.offset_from(meminfo.BaseAddress)) as usize)
 }
 
-enum ProcessDataKind {
-    CMDLINE,
-    CWD,
-    ENVIRON,
-}
-
 unsafe fn get_process_data(
     handle: HANDLE,
-    kind: ProcessDataKind,
+    ptr: LPVOID,
+    size: usize,
 ) -> Result<Vec<u16>, &'static str> {
+    let mut buffer: Vec<u16> = Vec::with_capacity(size / 2 + 1);
+    buffer.set_len(size / 2);
+    if ReadProcessMemory(
+        handle,
+        ptr as *mut _,
+        buffer.as_mut_ptr() as *mut _,
+        size,
+        std::ptr::null_mut(),
+    ) != TRUE
+    {
+        return Err("Unable to read process data");
+    }
+    Ok(buffer)
+}
+
+trait RtlUserProcessParameters {
+    fn get_cmdline(&self, handle: HANDLE) -> Result<Vec<u16>, &'static str>;
+    fn get_cwd(&self, handle: HANDLE) -> Result<Vec<u16>, &'static str>;
+    fn get_environ(&self, handle: HANDLE) -> Result<Vec<u16>, &'static str>;
+}
+
+macro_rules! impl_RtlUserProcessParameters {
+    ($t:ty) => {
+        impl RtlUserProcessParameters for $t {
+            fn get_cmdline(&self, handle: HANDLE) -> Result<Vec<u16>, &'static str> {
+                let ptr = self.CommandLine.Buffer;
+                let size = self.CommandLine.Length;
+                unsafe { get_process_data(handle, ptr as _, size as _) }
+            }
+            fn get_cwd(&self, handle: HANDLE) -> Result<Vec<u16>, &'static str> {
+                let ptr = self.CurrentDirectory.DosPath.Buffer;
+                let size = self.CurrentDirectory.DosPath.Length;
+                unsafe { get_process_data(handle, ptr as _, size as _) }
+            }
+            fn get_environ(&self, handle: HANDLE) -> Result<Vec<u16>, &'static str> {
+                let ptr = self.Environment;
+                unsafe {
+                    let size = get_region_size(handle, ptr as LPVOID)?;
+                    get_process_data(handle, ptr as _, size as _)
+                }
+            }
+        }
+    };
+}
+
+impl_RtlUserProcessParameters!(RTL_USER_PROCESS_PARAMETERS32);
+impl_RtlUserProcessParameters!(RTL_USER_PROCESS_PARAMETERS);
+
+unsafe fn get_process_params(handle: HANDLE) -> (Vec<String>, Vec<String>, PathBuf) {
+    #[inline]
+    fn default_values() -> (Vec<String>, Vec<String>, PathBuf) {
+        (Vec::new(), Vec::new(), PathBuf::new())
+    }
+
     if !cfg!(target_pointer_width = "64") {
-        return Err("Non 64 bit targets are not supported");
+        sysinfo_debug!("Non 64 bit targets are not supported");
```

Comment:
One small thing: I think it's a bit better to use `Result<>` as much as possible and leave `sysinfo_debug!` only for the last call. If for some reason we later want to use a different error reporting strategy, it would be easier. But this is just my opinion as a Rust beginner :)

Reply:
Good point! I'll keep the old `Err` then.